src/zen/ml/ZenTabsTidy.sys.mjs

```diff
@@ -0,0 +1,647 @@
+    return 0;
+  }
+
+  let dotProduct = 0;
+  let norm1 = 0;
+  let norm2 = 0;
+
+  for (let i = 0; i < vec1.length; i++) {
+    dotProduct += vec1[i] * vec2[i];
+    norm1 += vec1[i] * vec1[i];
+    norm2 += vec2[i] * vec2[i];
+  }
+
+  const denominator = Math.sqrt(norm1) * Math.sqrt(norm2);
+  return denominator === 0 ? 0 : dotProduct / denominator;
+}
+
+function calculateTextSimilarity(text1, text2) {
+  const words1 = new Set(
+    text1
+      .toLowerCase()
+      .split(/\s+/)
+      .filter((word) => word.length > 2)
+  );
+  const words2 = new Set(
+    text2
+      .toLowerCase()
+      .split(/\s+/)
+      .filter((word) => word.length > 2)
+  );
+
+  if (words1.size === 0 && words2.size === 0) {
+    return 1.0;
+  }
+  if (words1.size === 0 || words2.size === 0) {
+    return 0.0;
+  }
+
+  const intersection = new Set([...words1].filter((word) => words2.has(word)));
+  const union = new Set([...words1, ...words2]);
+  return intersection.size / union.size;
+}
+
+async function generateDynamicClusterLabel(clusterData) {
+  if (clusterData.length === 1) {
+    return generateClusterLabel(clusterData[0].domain);
+  }
+
+  try {
+    const label = await generateAiClusterLabel(clusterData);
+    if (label) {
+      return label;
+    }
+  } catch (error) {
+    console.warn('ZenTabsTidy: AI label generation failed, using fallback:', error);
+  }
+
+  const wordFreq = new Map();
+
+  for (const data of clusterData) {
+    const text = `${data.domain} ${data.title}`.toLowerCase();
+    const words = text.split(/\s+/).filter((word) => word.length > 3);
+
+    for (const word of words) {
+      wordFreq.set(word, (wordFreq.get(word) || 0) + 1);
+    }
+  }
+
+  let bestWord = null;
+  let bestCount = 0;
+
+  for (const [word, count] of wordFreq.entries()) {
+    if (count > 1 && count > bestCount) {
+      bestCount = count;
+      bestWord = word;
+    }
+  }
+
+  if (bestWord) {
+    return bestWord.charAt(0).toUpperCase() + bestWord.slice(1);
+  }
+
+  return generateClusterLabel(clusterData[0].domain);
+}
+
+async function generateAiClusterLabel(clusterData) {
+  if (!Services.prefs.getBoolPref('browser.ml.enable', false)) {
+    return null;
+  }
+
+  try {
+    const { createEngine } = ChromeUtils.importESModule(
+      'chrome://global/content/ml/EngineProcess.sys.mjs'
+    );
+
+    const tabDescriptions = clusterData
+      .map((data, idx) => {
+        let desc = `${idx + 1}. ${data.url}\n   Title: ${data.title}`;
+        if (data.opener) {
+          desc += `\n   ↳ Opened from: ${data.opener.domain} (${data.opener.title})`;
+        }
+        return desc;
+      })
+      .join('\n\n');
+
+    const titles = clusterData.map((data) => data.title).filter((title) => title.length > 0);
+    const keywords = extractKeywords(titles);
+    const domains = [...new Set(clusterData.map((data) => data.domain))].join(', ');
+
+    const prompt = `You are an expert organizer who creates concise, descriptive category names.
```

Comment:
Not sure if this is the best way we should do it, specially since there might be some possible vulnerability by directly injecting the scripts. Firefox's implementation (https://searchfox.org/firefox-main/source/browser/components/tabbrowser/SmartTabGrouping.sys.mjs) doesn't do this, for example